### PR TITLE
Support target identification in Game Logic

### PR DESCRIPTION
**Pull request by iche033**
Signed-off-by: Ian Chen <ichen@osrfoundation.org>

Updated GameLogicPlugin to supporting validating target reports. Teams have 3 attempts to ID a target vessel, target small object, and target large object. Time penalties are given for the first 2 incorrect attempts, and the run will terminate on 3rd incorrect attempt.

The targets are defined as parameters in the GameLogicPlugin:

```
      <target>
        <vessel>Vessel A</vessel>
        <small_object>small_box_a</small_object>
        <large_object>large_box_a</large_object>
      </target>
```

There can be multiple `<target>` SDF elements.

I added Ignition services for testing target Identification and validating time penalties. They will not be available in the actual competition run so I didn't add ign_ros bridges for these services or integration tests. Instructions for testing can be found in this wiki page:
https://github.com/osrf/mbzirc/wiki/Time-Penalties#target-identification-penalties

### CODE REVIEW

File: mbzirc_ign/CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,7 @@ find_package(ignition-fuel_tools7 REQUIRED)
 find_package(ignition-math6 REQUIRED)
 set(IGN_MATH_VER ${ignition-math6_VERSION_MAJOR})
 find_package(ignition-msgs8 REQUIRED)
+set(IGN_MSGS_VER ${ignition-msgs8_VERSION_MAJOR})
```
Comment (caguero) [inline mbzirc_ign/CMakeLists.txt:16]: Nit: Do we need this?

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 2]
```diff
@@ -790,6 +896,208 @@ void GameLogicPluginPrivate::Finish(const ignition::msgs::Time &_simTime)
   this->finished = true;
 }
 
+/////////////////////////////////////////////////
+bool GameLogicPluginPrivate::OnReportTargets(
+    const ignition::msgs::StringMsg_V &_req,
+    ignition::msgs::Boolean &_res)
+{
+  this->reports.push_back(_req);
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:904]: Should we use `reportMutex` here? I don't think it's used when reading its value either.

**Review by caguero — COMMENTED** [review]
It worked as expected. Just two minor comments in the code and one clarification:

* The tutorial says to report `small_object_a` as the correct small target object but the code is configured to be `small_box_a`. 
* The tutorial mentions to run the service to report the large target object three times in order to get the `finished` event. In my case, I only had to run it once to get into the `termination` state. I guess it's because I already did two incorrect reports (one for the the small target and one for the target vessel). Just checking if this is the expected behavior.

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 2 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc:905]: oops yeah that was meant to be done here. Added. [de1da17](https://github.com/osrf/mbzirc/pull/52/commits/de1da1726a12615716efd5b1d684e417c7ce8ea8)

File: mbzirc_ign/CMakeLists.txt (hunk 1 above)
Comment (iche033) [reply mbzirc_ign/CMakeLists.txt]: removed. [de1da17](https://github.com/osrf/mbzirc/pull/52/commits/de1da1726a12615716efd5b1d684e417c7ce8ea8)

**Comment by iche033** [comment]
> The tutorial says to report small_object_a as the correct small target object but the code is configured to be small_box_a

fixed

> The tutorial mentions to run the service to report the large target object three times in order to get the finished event. In my case, I only had to run it once to get into the termination state. I guess it's because I already did two incorrect reports (one for the the small target and one for the target vessel). Just checking if this is the expected behavior.

Found that the last service call instruction in the tutorial is incorrect. It reported the small target object instead of the large target object. I have fixed in the instruction. It should log the correct event now.

**Pull request merged**